open-craft/frontend-app-library-authoring
# [BD-14] [SE-2931] User Access Permissions added for Library.

This PR adds the ability to manage user permissions for librariesv2.

**Discussions**: https://docs.google.com/document/d/1hRrSmWWlGHhi-bh9AIlpSuLtJFnn2GpavNe8OQxthbE/edit

**Dependencies**: None

**Screenshots**: 
![Screen Shot 2020-09-04 at 10 42 20 AM](https://user-images.githubusercontent.com/1099483/92257907-97107080-ee9b-11ea-9ae7-cde291923ff5.png)
![Screen Shot 2020-09-04 at 10 42 41 AM](https://user-images.githubusercontent.com/1099483/92257919-9d9ee800-ee9b-11ea-83aa-9ff95f466164.png)
![Screen Shot 2020-09-04 at 10 42 57 AM](https://user-images.githubusercontent.com/1099483/92257925-a1326f00-ee9b-11ea-820a-566dc8826040.png)

**Merge deadline**: This PR is for code review only and is not to be merged.

**Testing instructions**:

1. Use the `edx-platform` branch `open-craft:fox/se-2931-user-access-adjustments` on the backend.
1. Follow the directions in https://github.com/edx/frontend-app-library-authoring/pull/4 up until item 15.
1. Create a new library of the 'complex' type.
1. In the top menu, there should be a 'User Access' link. Click it to visit the user access page.
1. Create and remove users, elevate their access, and depriv them. Verify the form resets upon add and that the user is added to the list.
1. Put bogus users and data into the email field to verify errors are handled well.
1. Delete users. Note, you should be able to delete yourself as a staff member, and still readd yourself, despite not having labeled privileges.
1. Verify that the interface does not let you remove or depriv yourself if there is no secondary admin.
1. Elevate a user who is not marked is_staff or is_superuser to 'staff' within the interface (that is, library author-- not marking them is_staff in the admin), and have them log in. Verify that the access control buttons are not present.


**Author notes and concerns**:

1. This is the largest amount of React code I've had to make so far, and the first time I've done real full component dev-- previously I modified other components. This is also my first time with Redux. I'd like feedback on how well I'm using those tools.
2. This task initially had an implementation similar to other pages currently in the repository. However, after speaking with @bradenmacdonald , he suggested using [the container pattern](https://www.freecodecamp.org/news/react-superpowers-container-pattern-20d664bdae65/) so that unit tests may be more easily written next sprint. This is my first time using this pattern and so I'd like his review to make sure I'm using/understanding it well.
3. It's been a while since I've done anything a11y related-- we're using paragon, which I believe is able to automate much of it, but I'd like to make sure I understand if I'm assuming its capabilities to be greater than they are in the code I'm writing, since I haven't immediately thought of any markup I'd have to customize here.

**Reviewers**
- [x] edX reviewer[s] TBD (Still need to have an arch level review here, yes?)
